File: nn/src/test/scala/com/intel/webscaleml/nn/nn/performTest/LinearPerform.scala
package com.intel.webscaleml.nn.nn.performTest

import java.io.{BufferedWriter, File, FileWriter}

import com.intel.webscaleml.nn.nn.Linear
import com.intel.webscaleml.nn.tensor.{RandomGenerator, torch}
import org.scalatest.{BeforeAndAfter, FlatSpec}
import org.scalatest.concurrent.Timeouts._
import org.scalatest.time.{Millis, Span}

/**
  * Created by yao on 5/19/16.
  */
class LinearPerform extends FlatSpec with BeforeAndAfter{
  val ForwardIterations = PerformTestData.forwardIterations
  val BackwardIterations = PerformTestData.backwardIterations
  val InputNum = 100
  val Seed = 100
  val allTestCases = List(
    TestCase(256 * 6 * 6, 4096, 55, 90),
    TestCase(4096, 4096,25, 55),
    TestCase(256 * 5 * 5, 128, 1.5, 2),
    TestCase(512, 512, 1.2, 2.5),
    TestCase(512, 10, 0.5, 0.5),
    TestCase(28 * 4 * 4, 768, 1, 2.5)
  )

  for (testCase <- allTestCases) {
    "A Linear" should s"forward with a good performance with parameters: ${testCase.inputSize}," +
      s" ${testCase.outputSize}" in {
      if (!TestUtils.isRun())
        cancel("Performance regression tests are forbidden")
      val input = torch.Tensor[Float](InputNum, testCase.inputSize).rand()
      val model = new Linear[Float](testCase.inputSize, testCase.outputSize)

      val timeMillis = TestUtils.testModuleForwardPerform(input, model, ForwardIterations)
      assert(timeMillis < testCase.ForwardTimeoutMillis)
    }
  }

  for (testCase <- allTestCases) {
    val model = new Linear[Float](testCase.inputSize, testCase.outputSize)
    val input = torch.Tensor[Float](InputNum, testCase.inputSize)
    val output = model.forward(input)
    val grad = output.clone().rand()

    it should s" backward with a good performance with parameters: ${testCase.inputSize}," +
      s" ${testCase.outputSize}" in {
      if (!TestUtils.isRun())
        cancel("Performance regression tests are forbidden")
      val timeMillis = TestUtils.testModuleBackwardPerform(input, grad, model, BackwardIterations)
      assert(timeMillis < testCase.BackwardTimeoutMillis)
    }

  }

  case class TestCase(inputSize: Int, outputSize: Int, ForwardTimeoutMillis: Double, BackwardTimeoutMillis: Double)
}

File: nn/src/test/scala/com/intel/webscaleml/nn/nn/performTest/LinearPerform.lua
--
-- Created by yansh on 16-6-22.
--

require "nn"

forwardIterations = 10
backwardIterations = 10
inputNum = 100
seed = 100
allTestCases = {
    {inputSize = 256 * 6 * 6, outputSize = 4096},--, 3000, 3900},
    {inputSize = 4096, outputSize = 4096},--1500, 2000},
    {inputSize = 256 * 5 * 5, outputSize = 128},-- 65, 90},
    {inputSize = 512, outputSize = 512},-- 21, 30},
    {inputSize = 512, outputSize = 10},-- 2, 2},
    {inputSize = 28 * 4 * 4, outputSize = 768}-- 28, 40}
}

torch.manualSeed(seed)
file = io.open("TorchPerform.csv", "a")
io.output(file)

for _, testCase in pairs(allTestCases) do
    local input = torch.Tensor(inputNum, testCase.inputSize)
    local model = nn.Linear(testCase.inputSize, testCase.outputSize)

    local timer = torch.Timer()
    for j = 1, forwardIterations do
        model:forward(input)
    end
    local timeSpend = timer:time().real / forwardIterations
    file:write("Linear(" .. testCase.inputSize .."," ..
            testCase.outputSize  .. ")forward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")
end

for _, testCase in pairs(allTestCases) do
    local model = nn.Linear(testCase.inputSize,testCase.outputSize)
    local input = torch.Tensor(inputNum, testCase.inputSize)
    local output = model:forward(input)
    local grads = output:clone()

    local timer = torch.Timer()
    for j = 1, backwardIterations do
        model:backward(input, grads)
    end
    local timeSpend = timer:time().real / backwardIterations

    file:write("Linear(" .. testCase.inputSize .."," ..
            testCase.outputSize  .. ")backward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")
end

file:close()
